### PR TITLE
Allow for variable declarations in readonly-generated code

### PR DESCRIPTION
**Pull request by Hamms**
When generating code directly from XML, we use a helper that initializes
a new readonly blockspace, populates it with blocks, and generates code
from there. We would formerly accomplish this by populating the user's
working blockspace with the blocks, generating the code, and then
replacing the user's original blockspace, but this had some problems.

Unfortunately, the code generator, when scanning the code for variables
whose declaration it wants to pull up into the top of the block, only
considers the blocks in the main block space when looking for those
variables. This can lead to problems, as with [course 4, stage 18](https://studio.code.org/s/course4/stage/18/puzzle/2)

The simple fix is to have the generator directly scan for variables in
the blocks that it's using.

### CODE REVIEW

File: blockly-core/generators/javascript.js [hunk 1]
```diff
@@ -108,8 +108,10 @@ Blockly.JavaScript.INFINITE_LOOP_TRAP = null;
 
 /**
  * Initialise the database of variable names.
+ * @param {Array.<Blockly.Block>} opt_blocks Optional blocks to query for
```
Comment (bcjordan) [inline blockly-core/generators/javascript.js]: Think you could bracket this like: `[opt_blocks]` to [mark it optional](http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values) in jsdoc

File: blockly-core/core/utils/variables.js [hunk 2]
```diff
@@ -42,13 +42,16 @@ Blockly.Variables.NAME_TYPE_LOCAL = 'LOCALVARIABLE';
 /**
  * Find all user-created variables.
  * Currently searches the main blockspace only
- * @param {Blockly.Block=} opt_block Optional root block.
+ * @param {Array.<Blockly.Block>} opt_blocks Optional root blocks.
```
Comment (bcjordan) [inline blockly-core/core/utils/variables.js]: May want to retain the `=` or bracket like `[opt_blocks]` to mark as jsdoc optional

File: blockly-core/core/utils/variables.js [hunk 3]
```diff
@@ -42,13 +42,16 @@ Blockly.Variables.NAME_TYPE_LOCAL = 'LOCALVARIABLE';
...
  * @return {!Array.<string>} Array of variable names.
  */
-Blockly.Variables.allVariables = function(opt_block) {
+Blockly.Variables.allVariables = function(opt_blocks) {
   var blocks;
-  if (opt_block) {
-    blocks = opt_block.getDescendants();
+  if (opt_blocks) {
+    opt_blocks = Array.isArray(opt_blocks) ? opt_blocks : [opt_blocks];
```
Comment (bcjordan) [inline blockly-core/core/utils/variables.js:51]: Looks like the parameter only currently gets used in python mode's `procedure_defreturn` so no real way to test, but seems fine to wrap it in any case.

File: blockly-core/core/utils/variables.js [hunk 4]
```diff
@@ -42,13 +42,16 @@ Blockly.Variables.NAME_TYPE_LOCAL = 'LOCALVARIABLE';
...
+    blocks = opt_blocks.reduce(function(blocks, block) {
```
Comment (bcjordan) [inline blockly-core/core/utils/variables.js:52]: Nice use of reduce!

**Comment by bcjordan** [comment]
LGTM

File: blockly-core/core/utils/variables.js (hunk 3 above)
Comment (Hamms) [inline blockly-core/core/utils/variables.js:51]: yeah, it should be logically equivalent

File: blockly-core/core/utils/variables.js (hunk 4 above)
Comment (Hamms) [inline blockly-core/core/utils/variables.js:52]: I love me a nice reduce. Just so elegant without straying _too_ far into unreadability

**Comment by Hamms** [comment]
(manually canceled test run since I forgot to [ci skip] my documentation change)

Thanks!

**Pull request merged**